File: 3.x.x/crates-utopia/dns/src/validator/dns.rs
use parking_lot::Mutex;
use serde_json::Value;
use utopia_validators::{Validator, ValueType};

use crate::client::Client;
use crate::message::{Question, Record};
use crate::Message;

/// PHP `Utopia\DNS\Validator\DNS`.
#[derive(Debug)]
pub struct DNS {
    target: String,
    type_code: u16,
    dns_server: String,
    records: Mutex<Vec<String>>,
    value: Mutex<String>,
    count: Mutex<usize>,
    reason: Mutex<String>,
}

impl DNS {
    const FAILURE_REASON_QUERY: &'static str = "DNS query failed.";
    const FAILURE_REASON_INTERNAL: &'static str = "Internal error occurred.";
    const FAILURE_REASON_UNKNOWN: &'static str = "";
    pub const DEFAULT_DNS_SERVER: &'static str = "8.8.8.8";

    #[must_use]
    pub fn new(target: impl Into<String>, type_code: u16, dns_server: impl Into<String>) -> Self {
        Self {
            target: target.into(),
            type_code,
            dns_server: dns_server.into(),
            records: Mutex::new(Vec::new()),
            value: Mutex::new(String::new()),
            count: Mutex::new(0),
            reason: Mutex::new(String::new()),
        }
    }

    pub fn records(&self) -> Vec<String> {
        self.records.lock().clone()
    }
}

impl Validator for DNS {
    fn description(&self) -> String {
        let reason = self.reason.lock().clone();
        if !reason.is_empty() && reason != "0" {
            return reason;
        }
        let type_verbose = Record::type_code_to_name(self.type_code)
            .map_or_else(|| self.type_code.to_string(), str::to_string);
        let mut messages = vec![format!(
            "DNS verification failed with resolver {}",
            self.dns_server
        )];
        let count = *self.count.lock();
        let value = self.value.lock().clone();
        let records = self.records.lock().join("', '");
        if count == 0 {
            messages.push(format!("Domain {value} is missing {type_verbose} record"));
            return format!("{}.", messages.join(". "));
        }
        let count_verbose = match count {
            1 => "one",
            2 => "two",
            3 => "three",
            4 => "four",
            5 => "five",
            6 => "six",
            7 => "seven",
            8 => "eight",
            9 => "nine",
            10 => "ten",
            n => {
                return {
                    messages.push(format!(
                        "Domain {value} has {n} incompatible {type_verbose} records: '{records}'"
                    ));
                    extra_caa(&mut messages, self.type_code);
                    format!("{}.", messages.join(". "))
                };
            }
        };
        if count == 1 {
            messages.push(format!(
                "Domain {value} has incorrect {type_verbose} value '{records}'"
            ));
        } else {
            messages.push(format!(
                "Domain {value} has {count_verbose} incompatible {type_verbose} records: '{records}'"
            ));
        }
        extra_caa(&mut messages, self.type_code);
        format!("{}.", messages.join(". "))
    }

    fn value_type(&self) -> ValueType {
        ValueType::String
    }

    fn is_valid(&self, value: &Value) -> bool {
        let Some(value) = value.as_str() else {
            *self.reason.lock() = Self::FAILURE_REASON_INTERNAL.into();
            return false;
        };
        *self.count.lock() = 0;
        *self.value.lock() = value.to_string();
        *self.reason.lock() = Self::FAILURE_REASON_UNKNOWN.into();
        self.records.lock().clear();

        let Ok(dns) = Client::new(&self.dns_server, 53, 5, false) else {
            *self.reason.lock() = Self::FAILURE_REASON_QUERY.into();
            return false;
        };
        let question = Question::new(value, self.type_code);
        let Ok(query_message) = Message::query(question, None, true) else {
            *self.reason.lock() = Self::FAILURE_REASON_QUERY.into();
            return false;
        };
        let Ok(response) = dns.query(&query_message) else {
            *self.reason.lock() = Self::FAILURE_REASON_QUERY.into();
            return false;
        };
        let query: Vec<&Record> = response
            .answers
            .iter()
            .filter(|r| r.type_code == self.type_code)
            .collect();
        *self.count.lock() = query.len();
        if query.is_empty() {
            if self.type_code == Record::TYPE_CAA {
                if let Ok(domain) = utopia_domains::Domain::new(value) {
                    if domain.get() == domain.get_apex() {
                        return true;
                    }
                    let mut parts: Vec<&str> = value.split('.').collect();
                    if parts.len() < 2 {
                        return false;
                    }
                    parts.remove(0);
                    let parent = parts.join(".");
                    let validator = Self::new(&self.target, self.type_code, &self.dns_server);
                    let result = validator.is_valid(&Value::String(parent));
                    *self.records.lock() = validator.records();
                    self.value.lock().clone_from(&validator.value.lock());
                    *self.count.lock() = *validator.count.lock();
                    self.reason.lock().clone_from(&validator.reason.lock());
                    return result;
                }
            }
            return false;
        }
        for record in query {
            if self.type_code == Record::TYPE_CAA {
                let extracted = extract_caa_domain(&record.rdata);
                self.records.lock().push(extracted.clone());
                if extracted == self.target {
                    return true;
                }
            } else {
                self.records.lock().push(record.rdata.clone());
            }
            if record.rdata == self.target {
                return true;
            }
        }
        false
    }
}

fn extra_caa(messages: &mut Vec<String>, type_code: u16) {
    if type_code == Record::TYPE_CAA {
        messages.push("Add new CAA record, or remove all other CAA records".into());
    }
}

fn extract_caa_domain(rdata: &str) -> String {
    let third = rdata.splitn(3, ' ').nth(2).unwrap_or("");
    let unquoted = third.trim_matches('"');
    unquoted.split(';').next().unwrap_or("").to_string()
}

File: packages/dns/src/DNS/Validator/DNS.php
<?php

namespace Utopia\DNS\Validator;

use Utopia\DNS\Client;
use Utopia\DNS\Message;
use Utopia\DNS\Message\Question;
use Utopia\DNS\Message\Record;
use Utopia\Domains\Domain;
use Utopia\Validator;

class DNS extends Validator
{
    protected const string FAILURE_REASON_QUERY = 'DNS query failed.';
    protected const string FAILURE_REASON_INTERNAL = 'Internal error occurred.';
    protected const string FAILURE_REASON_UNKNOWN = '';
    protected const string DEFAULT_DNS_SERVER = '8.8.8.8';

    // Memory from isValid to be used in getDescription
    /**
     * @var array<string>
     */
    public array $records = []; // Records found, even if wrong value
    public string $value = ''; // Domain being validated
    public int $count = 0; // Amount of records found
    public string $reason = ''; // Override for when reason is known ahead of time

    /**
     * @param string $target Expected value for the DNS record
     * @param int $type Type of DNS record to validate
     *  For value, use const from Record, such as Record::TYPE_A
     *  When using CAA type, you can provide exact match, or just issuer domain as $target
     * @param string $dnsServer DNS server IP or domain to use for validation
     */
    public function __construct(protected string $target, protected int $type = Record::TYPE_CNAME, protected string $dnsServer = self::DEFAULT_DNS_SERVER) {}

    public function getDescription(): string
    {
        if ($this->reason !== '' && $this->reason !== '0') {
            return $this->reason;
        }

        $typeVerbose = Record::typeCodeToName($this->type) ?? $this->type;

        $messages = [];

        $messages[] = "DNS verification failed with resolver {$this->dnsServer}";

        if ($this->count === 0) {
            $messages[] = 'Domain ' . $this->value . ' is missing ' . $typeVerbose . ' record';
            return implode('. ', $messages) . '.';
        }

        $records = implode("', '", $this->records);

        $countVerbose = match ($this->count) {
            1 => 'one',
            2 => 'two',
            3 => 'three',
            4 => 'four',
            5 => 'five',
            6 => 'six',
            7 => 'seven',
            8 => 'eight',
            9 => 'nine',
            10 => 'ten',
            default => $this->count
        };

        if ($this->count === 1) {
            $messages[] = "Domain {$this->value} has incorrect {$typeVerbose} value '{$records}'";
        } else {
            // Two or more
            $messages[] = "Domain {$this->value} has {$countVerbose} incompatible {$typeVerbose} records: '{$records}'";
        }

        if ($this->type === Record::TYPE_CAA) {
            $messages[] = 'Add new CAA record, or remove all other CAA records';
        }

        return implode('. ', $messages) . '.';
    }

    /**
     * Check if DNS record value matches specific value
     */
    public function isValid(mixed $value): bool
    {
        if (!\is_string($value)) {
            $this->reason = self::FAILURE_REASON_INTERNAL;
            return false;
        }

        $this->count = 0;
        $this->value = \strval($value);
        $this->reason = self::FAILURE_REASON_UNKNOWN;
        $this->records = [];

        $dns = new Client($this->dnsServer);

        try {
            $question = new Question($value, $this->type);
            $queryMessage = Message::query($question, recursionDesired: true);
            $response = $dns->query($queryMessage);
            $answers = $response->answers;

            // Some DNS servers return all records, not only type that's asked for
            // Likely occurs when no records of specific type are found
            $query = array_filter($answers, fn(\Utopia\DNS\Message\Record $record): bool => $record->type === $this->type);
        } catch (\Exception) {
            $this->reason = self::FAILURE_REASON_QUERY;
            return false;
        }

        $this->count = \count($query);

        if ($query === []) {
            // CAA records inherit from parent (custom CAA behaviour)
            if ($this->type === Record::TYPE_CAA) {
                $domain = new Domain($value);
                if ($domain->get() === $domain->getApex()) {
                    return true; // No CAA on apex domain means anyone can issue certificate
                }

                // Recursive validation by parent domain
                $parts = explode('.', $value);
                array_shift($parts);
                $parentDomain = implode('.', $parts);
                $validator = new self($this->target, $this->type, $this->dnsServer);
                $result = $validator->isValid($parentDomain);
                $this->records = $validator->records;
                $this->value = $validator->value;
                $this->count = $validator->count;
                $this->reason = $validator->reason;
                return $result;
            }

            return false;
        }

        foreach ($query as $record) {
            // CAA validation only needs to ensure domain
            if ($this->type === Record::TYPE_CAA) {
                // Extract domain; comments showcase extraction steps in most complex scenario
                $rdata = $record->rdata; // 255 issuewild "certainly.com;validationmethods=tls-alpn-01;retrytimeout=3600"
                $rdata = explode(' ', $rdata, 3)[2] ?? ''; // "certainly.com;validationmethods=tls-alpn-01;retrytimeout=3600"
                $rdata = trim($rdata, '"'); // certainly.com;validationmethods=tls-alpn-01;retrytimeout=3600
                $rdata = explode(';', $rdata, 2)[0]; // certainly.com

                $this->records[] = $rdata;
                if ($rdata === $this->target) {
                    return true;
                }
            } else {
                $this->records[] = $record->rdata;
            }

            if ($record->rdata === $this->target) {
                return true;
            }
        }

        return false;
    }

    /**
     * Is array
     *
     * Function will return true if object is array.
     */
    public function isArray(): bool
    {
        return false;
    }

    /**
     * Get Type
     *
     * Returns validator type.
     */
    public function getType(): string
    {
        return self::TYPE_STRING;
    }
}
